### ISSUE TITLE
Do not render ligatures under cursor (optionally?)

### ISSUE THREAD

**Issue by maximbaz** [kovidgoyal/kitty#461]
I'm playing with font ligatures using Fira Code, it is fun to see them rendered and they convey a meaning quite clearly, however it is sometimes confusing to actually _edit_ the rendered ligature.

For example, looking at this, the meaning is clear:

![image](https://user-images.githubusercontent.com/1177900/38778473-5655727a-40ba-11e8-8f0b-78063f007375.png)

However, what's not immediately obvious is what will happen if I press <kbd>Del</kbd> or <kbd>Backspace</kbd> in this case:

![image](https://user-images.githubusercontent.com/1177900/38778479-7a2a8afa-40ba-11e8-8fcc-d1fe96726250.png)

In fact, in this particular case the code is:

```javascript
if (a !== b) {
```

What I suggest to do is to **not render a ligature if a cursor is currently over it**, just render `!==`. 
Then it is very obvious what <kbd>Del</kbd> or <kbd>Backspace</kbd> will do 🙂 

I guess it should be relatively easy to achieve, because a ligature always takes the same number of cells as all of the the underlying "raw" characters.

I have seen people express this opinion for font creators (e.g. https://github.com/be5invis/Iosevka/issues/257), where they suggest to edit the font and choose a better design. However, I think rendering "raw" characters instead of a ligature when the cursor is above the glyph is a more generic and font-agnostic solution.

I'm not sure if this requires a configuration option or should become a new behavior for everyone.

----------------

UPDATE: [looks like](https://github.com/jwilm/alacritty/issues/50#issuecomment-277468592) I'm not the first one who thought of this (hehe), and Qterminal and Konsole already have such behavior:

![decomposition](https://cloud.githubusercontent.com/assets/20362502/22620945/a7e15214-eb0f-11e6-8f01-de7c4256f9b4.png)

**Comment by kovidgoyal** [comment]
Patches are welcome

**Comment by maximbaz** [comment]
Thanks, this is interesting. 

The first Konsole's issue [361547](https://bugs.kde.org/show_bug.cgi?id=361547) will not be a problem in kitty due to the way how ligatures are implemented, the characters will never move because by design the "ligaturized" text takes exactly the same number of cells as the "raw" text.

![ligature_size](https://user-images.githubusercontent.com/1177900/38783847-6ea95a32-4108-11e8-8d84-93d843528f6b.gif)

Very interesting observation with colors 🙂

> Should the cursor only prevent ligatures across its left edge, or across both edges? Or should it depend on the cursor's shape (i-beam vs. rectangle)?

Hm, very good question. 

My first thought was that with rectangle and underline cursors I care to see the "raw characters" only when the cursor is directly over some part of the operator. The beam cursor usually indicates an insert mode where I can press <kbd>Del</kbd> or <kbd>Backspace</kbd> (so I also care about the previous and the next character comparing to my cursor's position). However, the cursor shape is not a true indicator of the "mode", for example I just opened `nano`, I see rectangle cursor and yet I can press <kbd>Del</kbd> and <kbd>Backspace</kbd>.

So, in summary, I think regardless of the cursor shape, the ligatures should not be rendered when the cursor is touching the operator with any of its sides.

This gif shows all of the cursor positions where I think the operator should be rendered as `>=` instead of the ligature:

![ligature_place](https://user-images.githubusercontent.com/1177900/38783863-c5c92f90-4108-11e8-9d41-04feb784b3bc.gif)

**Comment by kovidgoyal** [comment]
It depends very much on how fonts do the ligatures.You can have fonts that convert m codepoints to n glyphs where n can be one. So in general it is not possible to detect parts of ligatures corresponding to code-points. kitty does it crudely by simply splitting the ligature into cell blocks and assuming each block corresponds to the codepoint, as you can see with differently colored codepoints. But this is, in general not true, take for example !== becoming ≢ there is no real sense in which some part of that ligature corresponds to ! and the rest to ==. 

I'm actually not a fan of the cursor preventing ligatures. After all, I find symbols changing on the screen when moving the cursor distracting, and the cursor moves a lot more than it edits (in editors as opposed to the command line). Not to mention that if this is to be implemented it means you have to re-layout potentially two entire lines every time the cursor moves. That said, I wont refuse a PR to implement an option for this.

**Comment by egmontkob** [comment]
I'm not a fan of ligatures in terminal emulators at all :) ([VTE bug 762832 comment 1](https://bugzilla.gnome.org/show_bug.cgi?id=762832#c1))

**Comment by kovidgoyal** [comment]
Ligatures that span end of line are not a problem in my experience. They are simply rendered as normal characters. I haven't seen a single bug report complaining about that in kitty. This happens automatically if you do cell based rendering. You do have to jump through a lot of hoops to make harfbuzz work with cell based rendering, but once you do that, there are no major issues.

**Comment by hendrysadrak** [comment]
For reference how WebStorm does it.
```
asd: () => {
```
![asd1](https://user-images.githubusercontent.com/4000963/38855295-c1f25e90-422b-11e8-80e7-503c057a691f.gif)
```
if ( a >= b ) {
```
![asd2](https://user-images.githubusercontent.com/4000963/38855374-f6810e72-422b-11e8-9fc0-e9c0faaf39e9.gif)

**Comment by kovidgoyal** [comment]
I'm confused, I dont see any ligatures changing in your screencaps?

**Comment by hendrysadrak** [comment]
@kovidgoyal Yeah, just an example how webstorm is handling this situation. There it doesn't change, shows ligatures as they are. First one is two letters long, second one is one letter long.

**Comment by maximbaz** [comment]
The current behavior in kitty is exactly the same, and this issue describes the imperfections of such behavior. Please read the description 😉

**Comment by kovidgoyal** [comment]
It's complex. You'd need to get a good grasp of the way text rendering works in kitty. Every cell is represented by two structures, the CPUCell and the GPUCell. The CPUCell contains the text, the GPUCell contains the index into the texture map for the glyph corresponding to the text in the cell. So you would need to. The GPUCell is updated whenever any text on that line changes. You would then need to update it whenever the cursor changes position as well. 

Then you would need to convince harfbuzz to render the text without ligatures when the cursor is over it and store that in the texture map separately. and change the indices in the GPUCell to point to it.

The relevant code is mostly in shaders.c and fonts.c

**Comment by Luflosi** [comment]
Do you know how to make harfbuzz render without the ligatures? Would rendering each letter individually work? Is there a better alternative?

**Comment by kovidgoyal** [comment]
Nope sorry, I'm not really a harfbuzz expert, whatever I know about it, I learned on the job, so to speak :)

I vaguely recall that harbuzz has the ability to turn on and off opentype features, maybe turning off hte liga feature will do the trick, but I'm not sure.

**Comment by Luflosi** [comment]
Ok, here is my (not clean) test code in `fonts.c` that manages to turn off ligatures for FiraCode for anyone interested:
```c
    int nFeatures = 0;
    hb_feature_t* features = NULL;
    hb_tag_t tag;

    tag = hb_tag_from_string("liga", strlen("liga"));
    features = realloc(features, (nFeatures + 1) * sizeof(hb_feature_t));
    features[nFeatures].tag = tag;
    features[nFeatures].start = 0;
    features[nFeatures].end = (unsigned int) -1;
    features[nFeatures].value = 0;
    nFeatures++;

    tag = hb_tag_from_string("rlig", strlen("rlig"));
    features = realloc(features, (nFeatures + 1) * sizeof(hb_feature_t));
    features[nFeatures].tag = tag;
    features[nFeatures].start = 0;
    features[nFeatures].end = (unsigned int) -1;
    features[nFeatures].value = 0;
    nFeatures++;

    tag = hb_tag_from_string("dlig", strlen("dlig"));
    features = realloc(features, (nFeatures + 1) * sizeof(hb_feature_t));
    features[nFeatures].tag = tag;
    features[nFeatures].start = 0;
    features[nFeatures].end = (unsigned int) -1;
    features[nFeatures].value = 0;
    nFeatures++;

    tag = hb_tag_from_string("calt", strlen("calt"));
    features = realloc(features, (nFeatures + 1) * sizeof(hb_feature_t));
    features[nFeatures].tag = tag;
    features[nFeatures].start = 0;
    features[nFeatures].end = (unsigned int) -1;
    features[nFeatures].value = 0;
    nFeatures++;

    hb_shape(font, harfbuzz_buffer, features, nFeatures);
```
The final code may only want to turn off the `calt` feature as the other three don't do anything for FiraCode.
I may continue this some other time, in the meantime anyone is free to try to make progress on this.

**Comment by Luflosi** [comment]
@maximbaz did you see my PR yet? Feel free to try it out yourself.

**Comment by blueyed** [comment]
Awesome, thanks @Luflosi, and @kovidgoyal, of course!

I will give this a try for a while - I can imagine this to be useful to be changed on-the-fly, i.e. via a shortcut to change the config during runtime.

**Comment by Luflosi** [comment]
Would something like #1292 do what you want or would you like a keyboard shortcut specifically for changing this option? Maybe there are other ways to implement this too, e.g. remote control.
@kovidgoyal what would be the best option if we were to implement something like this?

**Comment by kovidgoyal** [comment]
Let's not jump the gun. Let @blueyed first come back with
whether a way to change it dynamically is useful or not, and if so, in
what use case.

**Comment by blueyed** [comment]
Sure, it was just a quick thought.

Some feedback already: with something like "win->w_p_pvw" the "->" ligature is only displayed when after the whole word/expression - I would expect it to be displayed as a ligature already if after the "->w".  I.e. with "foo->b_ar" (cursor at "_") it should render the ligature already, since the ligature is only "->", not the whole word(s).

**Comment by Luflosi** [comment]
If I understand what you mean correctly, then this is exactly how it behaves for me:
![screenshot](https://user-images.githubusercontent.com/15217907/54700070-8a2a8080-4b32-11e9-938a-de0a9c7ab5a9.png)
![screenshot](https://user-images.githubusercontent.com/15217907/54700081-8eef3480-4b32-11e9-9ec8-02e0bf49c9bd.png)
Is it different for you?

**Comment by kovidgoyal** [comment]
@Luflosi if I type ->a i get no ligature, if I type -><space> I do get a ligature, this is with

kitty -o font_family=Fira\ Code -o font_size=16 --config NONE -o disable_ligatures_under_cursor=y sh -c read

**Comment by blueyed** [comment]
@Luflosi 
That's how I would expect it, but I seem to only get it after whitespace.
Using `kitty -o font_family=Iosevka\ Term -o font_size=16 --config NONE -o disable_ligatures_under_cursor=y sh -c read` (Iosevka via https://github.com/be5invis/Iosevka/issues/248#issuecomment-414131620).

**Comment by Luflosi** [comment]
@kovidgoyal I can reproduce the issue, very interesting. Seems like there is more work to be done to get this working properly. Don't count on me til next week though, I'm a little busy.

**Comment by kovidgoyal** [comment]
@Luflosi you have to break up the run at the transition point when the value of disable_ligature changes.

**Comment by kovidgoyal** [comment]
In fact to make this work robustly, the only way I can see is to RENDER the full line, then check if any ligatures on the line intersect with the cursor position, if they do, then re-render just the cells in that ligature. Which means you will need to work at the level of  the Group. For a multi-codepoint group check if it intersects with the cursor and split it up if so. Going to be a fair bit of work.

**Comment by kovidgoyal** [comment]
I have implemented it using glyph groups. Not really tested it a lot. I leave testing/fine-tuning to you.

**Comment by Luflosi** [comment]
Awesome. I'll use it for a while and see if I can find any issues.

**Comment by blueyed** [comment]
I've noticed some delays/hangs with writing lately (might have started with the last update of this, but am not sure), where e.g. the shell prompt does not appear, until some redrawing is triggered (e.g. pressing a key, but also moving to another tmux pane).
Have not really investigated (and it is not reproducible then by just doing the same action / output again), but thought that it might be related to this feature.

**Comment by kovidgoyal** [comment]
No, it's more likely due to the refactoring to move the event loop into
GLFW. It's likely that some of the timer/event code is slightly buggy, I
just haven't had the time to investigate it. It's why I am not releasing
0.14 until I have a chance to investigate.

**Comment by josswright** [comment]
As this code is already written, would it be realistic to extend this to an option to allow disabling ligatures entirely? I'm seeing more and more fonts that are otherwise very nicely designed including them, and they''re often more annoying than useful.

**Comment by kovidgoyal** [comment]
This code is not really suitable for turning off all ligatures. For that, the correct way to do it would be to disable the liga and calt features in harfbuzz at startup, unconditionally.

**Comment by kovidgoyal** [comment]
See https://github.com/kovidgoyal/kitty/commit/934336e64278dae88ab87aa75c9bf553a5931fa6

**Comment by josswright** [comment]
Amazing response, as always!

Oddly enough, this seems to work as intended for the ligatures I've tried in Hasklig and for some ligatures in Space Mono, but not for others. For reference, maths characters `<-` and `->` are correctly disabled in Space Mono, but the `fi` and `fl` ligatures are still enabled (and not rendering correctly).

If there's something I can do to test, please do let me know!

**Comment by kovidgoyal** [comment]
It will be because those ligatures are not using the calt OpenType
feature but some other table for it. IMO a monospace font really has no
business using fi ligatures anyway. The whole point of the fi ligature
was to save space while maintaining legibility, which does not apply to
monospace fonts.

You could probably fix it by disabling more OpenType features, in
addition to calt such as liga and maybe kern.

**Comment by josswright** [comment]
Looking around it more, it looks that lots of people are having this issue with Space Mono as it's designed as a monospace display font, not a coding font. I did find someone who made a version with ligatures removed [Space Mono NL](http://tweetcompressor.com/temp/spacemono-nl.zip), so that's probably the best solution here rather than fighting a very non-standard font.

Great to have this ligature-disabling option built into kitty, though. Thanks again!
